New CRM for MMF (PAM)

whannah1 opened the pull request:
Integration of new CRM for MMF, referred to as the "portable atmosphere model" or PAM, that uses the P3 and SHOC physics implementation from SCREAM/EAMxx. The new CRM is used by the compsets with "MMF2" in their name. 

While the PAM code is isolated to a submodule, several large changes outside were required to facilitate this integration. The  PAM interface code is a substantial part of this, but there was also new conflicts created by EAMv3 integration efforts which led to the creation of various "dummy" modules to avoid building the conflicting modules, which are unnecessary for MMF.

A new integration is added to test the new CRM => SMS_Ln3.ne4pg2_ne4pg2.F2010-MMF2

A few things to note about the current MMF2 implementation
- When running with PAM on the GPU the use of openMP threads is problematic for reasons that are unclear. This seems fixable, but more work is needed to look into this. For now, running on GPUs without OMP threads is sufficient, and CPU cases are not affected by this problem
- Although some offline PAM tests have used longer time steps, when running within the MMF configuration at ne30 the PAM time step needs to be 4 seconds for stability. A lot of performance can be gained by increasing this time step, but more work is needed to understand the source of the instability. 
- PAM performance also seems highly machine dependent. For example, using 22 nodes on Perlmutter runs 6x faster than 64 nodes on Summit for the same case.

[BFB]

singhbalwinder reviewed (approved):
Looks good! I just left a very minor comment.

whannah1 commented on components/eam/bld/build-namelist:
```diff
@@ -1705,13 +1705,15 @@ if ( !($phys eq 'ideal' or $phys eq 'adiabatic') ) {
+        # } else {
+        #     add_default($nl, 'liqcldoptics', 'val' => 'gammadist');
+        #     add_default($nl, 'icecldoptics', 'val' => 'mitchell');
+        # }
```
yea, I'll do that. I left them there because I assume we'll switch to those optics schemes when @brhillman has a chance to work on this, but it'll be easy enough to add those lines when we get to that point.

whannah1 replied:
done.

rljacob commented on cime_config/machines/cmake_macros/gnugpu_summit.cmake:
```diff
@@ -20,3 +20,5 @@ set(NETCDF_C_PATH "$ENV{NETCDF_C_PATH}")
 set(NETCDF_FORTRAN_PATH "$ENV{NETCDF_FORTRAN_PATH}")
 set(PNETCDF_PATH "$ENV{PNETCDF_PATH}")
 set(USE_CUDA "TRUE")
+string(APPEND KOKKOS_OPTIONS " -DKokkos_ARCH_POWER9=On -DKokkos_ARCH_VOLTA70=On -DKokkos_ENABLE_CUDA=On -DKokkos_ENABLE_CUDA_LAMBDA=On")
```
Why is PAM/MMF modifying Kokkos opts?

xyuan commented:
For summit GPU runs, it needs to build Kokkos with GPU support, the default Kokkos build doesn’t have GPU support.

From: Robert Jacob ***@***.***>
Date: Tuesday, September 12, 2023 at 10:36 AM
To: E3SM-Project/E3SM ***@***.***>
Cc: Yuan, Xingqiu ***@***.***>, Review requested ***@***.***>
Subject: Re: [E3SM-Project/E3SM] New CRM for MMF (PAM) (PR #5906)

@rljacob commented on this pull request.

________________________________

In cime_config/machines/cmake_macros/gnugpu_summit.cmake<https://github.com/E3SM-Project/E3SM/pull/5906#discussion_r1323144786>:

> @@ -20,3 +20,5 @@ set(NETCDF_C_PATH "$ENV{NETCDF_C_PATH}")

 set(NETCDF_FORTRAN_PATH "$ENV{NETCDF_FORTRAN_PATH}")

 set(PNETCDF_PATH "$ENV{PNETCDF_PATH}")

 set(USE_CUDA "TRUE")

+string(APPEND KOKKOS_OPTIONS " -DKokkos_ARCH_POWER9=On -DKokkos_ARCH_VOLTA70=On -DKokkos_ENABLE_CUDA=On -DKokkos_ENABLE_CUDA_LAMBDA=On")

Why is PAM/MMF modifying Kokkos opts?

—
Reply to this email directly, view it on GitHub<https://github.com/E3SM-Project/E3SM/pull/5906#pullrequestreview-1622416476>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAFCG2DZY7VTBPFVMBGGUDLX2BXO5ANCNFSM6AAAAAA4GWU26A>.
You are receiving this because your review was requested.Message ID: ***@***.***>

rljacob commented on components/eam/src/physics/crm/dummy_modules/debug_info.F90:
```diff
@@ -0,0 +1,5 @@
+!-------------------------------------------------------------------------------
+! Dummy module to override src/physics/cam/debug_info.F90
+!-------------------------------------------------------------------------------
+end module debug_info
```
Why is github showing a red circle with a line through it at the end of the file?  Is it missing a newline at the end of the last line?

whannah1 replied:
yea, missing newline, I'll fix that.

jlee1046 commented on components/eam/bld/config_files/definition.xml:
```diff
@@ -348,12 +348,15 @@ Switch to enable or disable ECPP in CAM: 0=off, 1=on.
+<entry id="crm" valid_values="sam,samomp,samxx,pam" value="sam">
 MMF CRM model
 </entry>
+<entry id="pam_dycor" valid_values="awfl,spam" value="spam">
```
Is there a comment or documentation somewhere that tells difference between awfl and spam ?

whannah1 replied:
No, but I should definitely add something somewhere. I don't really intend for this to be something that anyone would change, but I still think it will be useful to compare the two dycors even if we only really use one. We don't seem to have any good documentation of the dycor options in the PAM repo, so I'll add a comment in this file to clarify the difference.

jlee1046 commented on components/eam/bld/configure:
```diff
@@ -358,9 +362,10 @@ GetOptions(
     "crm_nx_rad=s"              => \$opts{'crm_nx_rad'},
     "crm_ny_rad=s"              => \$opts{'crm_ny_rad'},
-    "MMF_microphysics_scheme=s"     => \$opts{'MMF_microphysics_scheme'},
+    "MMF_microphysics_scheme=s" => \$opts{'MMF_microphysics_scheme'},
```
what's '=s' for?

singhbalwinder replied:
In Perl, it means that this argument expects a string argument.

jlee1046 commented on components/eam/bld/configure:
```diff
@@ -1812,12 +1818,18 @@ if (defined $opts{'use_MMF'}) {
     if (defined $opts{'crm_adv'}) { $cfg_cppdefs .= " -D_$crm_adv "; }
-    $cfg_cppdefs .= " -D$MMF_microphysics_scheme -DYES3DVAL=$yes3Dval";
+    $cfg_cppdefs .= " -DYES3DVAL=$yes3Dval";
+    if ($MMF_microphysics_scheme eq 'sam1mom') {
```
this maybe a style difference so you may leave it as it is. I would combine if & elsif statement into one since $cfg_cppdefs .= "-D$MMF_microphysics_scheme" is the same.

jlee1046 commented on components/eam/bld/configure:
```diff
@@ -2660,10 +2676,10 @@ sub write_filepath
           }
                               
-        } elsif ($crm eq 'samxx') {
+        } elsif (($crm eq 'samxx') || ($crm eq 'pam')) {
```
is this intended to be an empty in elsif body?

whannah1 replied:
yes, just to hold the comment about not needing this section. I guess it's kinda silly to add the else if just for this though. Not sure why it was done this way originally. I'll change it.

whannah1 replied:
done

jlee1046 commented on components/eam/bld/config_files/definition.xml:
```diff
@@ -348,12 +348,15 @@ Switch to enable or disable ECPP in CAM: 0=off, 1=on.
 Switch to enable or disable fractional cloudiness in m2005 microphysics in CRM: 0=off, 1=on.
 </entry>
-<entry id="MMF_microphysics_scheme" valid_values="sam1mom,m2005" value="sam1mom">
+<entry id="MMF_microphysics_scheme" valid_values="sam1mom,m2005,p3" value="sam1mom">
```
m2005 is not valid option. Don't you want to delete it?

whannah1 replied:
yea, I'll do that.

whannah1 replied:
done

jlee1046 commented on components/eam/cime_config/config_component.xml:
```diff
@@ -68,12 +68,14 @@
+      <value compset="_EAM%.*MMF"      >-use_MMF -nlev 60 -crm_nz 50</value>
+      <value compset="_EAM%.*MMF"      >-crm_dx 2000 -crm_nx 64 -crm_ny 1 </value>
       <value compset="_EAM%.*MMF1"     >-MMF_microphysics_scheme sam1mom -chem none</value>
-      <value compset="_EAM%.*MMF"      >-use_MMF_VT</value>   <!-- CRM variance transport -->
```
do use_MMF_VT and use_MMF_ESMT need to be set in the cam_config_option additionally when using MMF1?

whannah1 replied:
Good catch. It actually doesn't matter right now since it's not implemented in PAM yet, but it's worth disabling it for PAM cases so I'll update this line.

whannah1 replied:
FYI re-enabled VT and ESMT only for MMF1/SAM option

jlee1046 commented on components/eam/cime_config/config_component.xml:
```diff
@@ -141,10 +143,10 @@
+      <!-- TODO: revisit the MMF2 use case when aerosols are not prescribed -->
+      <value compset="20TR(?:SOI)?_EAM%.*MMF"           >20TR_MMF-1mom_CMIP6</value>
+      <value compset="1950(?:SOI)?_EAM%.*MMF"           >1950_MMF-1mom_CMIP6</value>
+      <value compset="2010(?:SOI)?_EAM%.*MMF"           >2010_mmf_1mom</value>
```
any reason to have mmf in 2010_mmf_1mom as lowercase?

jlee1046 commented on components/eam/src/physics/crm/crm_physics.F90:
```diff
@@ -323,12 +341,12 @@ subroutine crm_physics_init(state, pbuf2d, species_class)
-   use constituents,   only: apcnst, bpcnst, cnst_name, cnst_longname, cnst_get_ind
+   use cam_history,           only: addfld, hist_fld_active
+   use constituents,          only: apcnst, bpcnst, cnst_name, cnst_longname, cnst_get_ind
+   use constituents,          only: pcnst, cnst_get_ind
```
Doesn't this line overwrite line 345?

whannah1 replied:
yea, we still access all the same module variables, so this is ok.

rljacob commented:
@jlee1046 Have your concerns been addressed?

jlee1046 commented:
> @jlee1046 Have your concerns been addressed?

I was on vacation and haven't finished going through all the changes. I plan on finish reviewing in a day or two.

jlee1046 commented on components/eam/src/physics/crm/pam/pam_debug.h:
```diff
@@ -0,0 +1,211 @@
+      rho_i(k,j,i,n)
+    );
+  });
+}
```
You may want to add a new line at the end of the code to get rid of that red circle with a bar

whannah1 replied:
done.

jlee1046 reviewed (approved):
All look good to me.

whannah1 commented:
I pushed some major recent updates that appear to have addressed many of the problems I have been seeing in global simulations. Here's a summary of the major things I changed:

- fixed a problem with the conversion between temperature and potential temperature in the PAM-C dycor which inadvertently imposed the water vapor forcing into the CRM temperature response - I think this is the most important change
- Added a hyperdiffusion calculation to account for the missing horizontal turbulence mixing - hopefully we can replace this such that SHOC handles the horizontal mixing
- Switched to gradually forcing the dry density instead of instantaneously reseting it at the beginning of each CRM call
- Switched from using adding the total water forcing to the vapor field and adjusting the temperature to account for the phase conversion
- Switched the water species to force the mixing ratio and convert back to density to be consistent with how the GCM feedback tendencies are formulated and, additionally, account for gradually forced dry density
- Added forcing for hydrometeor number concentrations

brhillman commented:
Merged to next.
